webapp/javascript/components/TimelineChart/Annotations.plugin.tsx

```diff
@@ -0,0 +1,285 @@
+  ctx.moveTo(x + radius, y);
+  ctx.lineTo(r - radius, y);
+  ctx.quadraticCurveTo(r, y, r, y + radius);
+  ctx.lineTo(r, y + h - radius);
+  ctx.quadraticCurveTo(r, b, r - radius, b);
+  ctx.lineTo(x + radius, b);
+  ctx.quadraticCurveTo(x, b, x, b - radius);
+  ctx.lineTo(x, y + radius);
+  ctx.quadraticCurveTo(x, y, x + radius, y);
+  ctx.fill();
+  ctx.stroke();
+}
+
+(function ($) {
+  function init(plot: jquery.flot.plot & jquery.flot.plotOptions) {
+    const annotationsPositions: AnnotationPosition[] = [];
+    const placeholder = plot.getPlaceholder();
+
+    function onHover(_: unknown, pos: { x: number; y: number }) {
+      if (annotationsPositions?.length) {
+        const { x, y } = getCursorPositionInPx(plot, pos);
+
+        const annotation = findAnnotationByCursorPosition(
+          x,
+          y,
+          annotationsPositions
+        );
+
+        if (annotation) {
+          placeholder.trigger('hoveringOnAnnotation', [{ hovering: true }]);
+        } else {
+          placeholder.trigger('hoveringOnAnnotation', [{ hovering: false }]);
+        }
+      }
+    }
+
+    function onClick(
+      _: unknown,
+      pos: { x: number; pageX: number; pageY: number; y: number }
+    ) {
+      const options: IFlotOptions = plot.getOptions();
+      const container = inject($);
+      const containerEl = container?.[0];
+
+      ReactDOM.unmountComponentAtNode(containerEl);
+
+      const ContextMenu = options?.ContextMenu;
+
+      const { x, y } = getCursorPositionInPx(plot, pos);
+
+      const annotation = findAnnotationByCursorPosition(
+        x,
+        y,
+        annotationsPositions
+      );
+
+      if (ContextMenu && containerEl) {
+        const timestamp = Math.round(pos.x / 1000);
+        const value = annotation
+          ? {
+              timestamp: annotation.timestamp,
+              content: annotation.content,
+            }
+          : null;
+
+        ReactDOM.render(
+          <ContextMenu
+            click={{ ...pos }}
+            containerEl={containerEl}
+            timestamp={timestamp}
+            value={value}
+          />,
+          containerEl
+        );
+      }
+    }
+
+    plot.hooks!.draw!.push((plot, ctx: CtxType) => {
+      const o: IFlotOptions = plot.getOptions();
+
+      if (o.annotations?.length) {
+        const axes = plot.getAxes();
+        const plotOffset: { top: number; left: number } = plot.getPlotOffset();
+        const extractedX = extractRange(plot, axes, 'x');
+        const extractedY = extractRange(plot, axes, 'y');
+
+        o.annotations.forEach((a: AnnotationType) => {
+          const left: number =
+            Math.floor(extractedX.axis.p2c(a.timestamp * 1000)) +
+            plotOffset.left;
+          const yMax =
+            Math.floor(extractedY.axis.p2c(extractedY.axis.min)) +
+            plotOffset.top;
+          const yMin = 0 + plotOffset.top;
+          const lineWidth = 1;
+          const subPixel = lineWidth / 2 || 0;
```

Comment:
what is a subpixel? what is this necessary?

Reply:
Honestly I got this in built-in flot rendering engine. As I understand flot uses this approach for accurate positioning some elements on X axis (+/-0.5px).